src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs

```diff
@@ -73,122 +73,260 @@ public override bool WillEncode(int unicodeScalar)
             return NeedsEscaping((char)unicodeScalar);
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public override unsafe int FindFirstCharacterToEncode(char* text, int textLength)
         {
             if (text == null)
             {
                 throw new ArgumentNullException(nameof(text));
             }
 
+            Debug.Assert(textLength >= 0);
+
+            if (textLength == 0)
+            {
+                goto AllAllowed;
+            }
+
             int idx = 0;
+            short* ptr = (short*)text;
+            short* end = ptr + (uint)textLength;
 
 #if NETCOREAPP
-            if (Sse2.IsSupported)
+            if (Sse2.IsSupported && textLength >= Vector128<short>.Count)
+            {
+                goto VectorizedEntry;
+            }
+
+        Sequential:
+#endif
+            Debug.Assert(textLength > 0 && ptr < end);
+
+            do
+            {
+                Debug.Assert(text <= ptr && ptr < (text + textLength));
+
+                if (NeedsEscaping(*(char*)ptr))
+                {
+                    goto Return;
+                }
+
+                ptr++;
+                idx++;
+            }
+            while (ptr < end);
+
+        AllAllowed:
+            idx = -1;
+
+        Return:
+            return idx;
+
+#if NETCOREAPP
+        VectorizedEntry:
+            short* vectorizedEnd = end - 2 * Vector128<short>.Count;
```

Comment:
Potential underflow. The if check on line 95 validates that we have 8 chars we can read, but this subtraction assumes we're going to read 16 chars at a time.

In reality this shouldn't be a problem because the lowermost addresses of memory should be off-limits, but we should still avoid such underflows in the code.

Reply:
Good catch.